File: optionals/med-4.0.0/tests/c/2.3.6/test28.c
/*  This file is part of MED.
 *
 *  COPYRIGHT (C) 1999 - 2019  EDF R&D, CEA/DEN
 *  MED is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  MED is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with MED.  If not, see <http://www.gnu.org/licenses/>.
 */

/******************************************************************************
 * - Nom du fichier : test28.c
 *
 * - Description : lecture d'un maillage structure (grille cartesienne |
 *                 grille polaire) dans le fichier test27.med
 *
 *****************************************************************************/

#include <med.h>
#define MESGERR 1
#include "med_utils.h"
#include <string.h>

#ifdef DEF_LECT_ECR
#define MODE_ACCES MED_LECTURE_ECRITURE
#elif DEF_LECT_AJOUT
#define MODE_ACCES MED_LECTURE_AJOUT
#else
#define MODE_ACCES MED_CREATION
#endif

int main (int argc, char **argv)


{
  int i;
  med_err ret = 0;
  med_idt fid;
  /* la dimension du maillage */
  med_int mdim,axe,j;
  med_float *indices;
  med_int nind,nmaa;
  /* nom du maillage de longueur maxi MED_TAILLE_NOM */
  char maa[MED_TAILLE_NOM+1];
  /* composantes et unites */
  char comp[MED_TAILLE_PNOM+1];
  char unit[MED_TAILLE_PNOM+1];
  char desc[MED_TAILLE_DESC+1];
  med_maillage type_maillage;
  med_type_grille type;
  med_table quoi;
  med_int nnoeuds;
  char *comp2, *unit2;
  med_float *coo;
  med_int *structure_grille;
  med_repere repere;

  /* Ouverture du fichier test17.med en lecture seule */
  fid = MEDouvrir("test27.med",MED_LECTURE);
  if (fid < 0) {
    MESSAGE("Erreur a l'ouverture du fichier test27.med");
    return -1;
  }   

  /* Lecture du nombre de maillage */
  nmaa = MEDnMaa(fid);
  if (nmaa < 0) {
    MESSAGE("Erreur a la lecture du nombre de maillage");
    return -1;
  }   

  /* On boucle sur les maillages et on ne lit que la grille cartesienne 
     et la grille desturcutree */
  for (i=0;i<nmaa;i++) {
    
    /* On repere le maillage qui nous interesse */
    if (MEDmaaInfo(fid,i+1,maa,&mdim,&type_maillage,desc) < 0) {
      MESSAGE("Erreur a la lecture des infos sur le maillage");
      return -1;
    }   
    printf("Maillage de nom [%s] de dimension "IFORMAT" \n",maa,mdim);
    if (type_maillage == MED_STRUCTURE)
      printf("- Type : Maillage structure \n");
    else
      printf("- Type : Maillage non structure \n");
    
    /* On regarde le type de la grille */
    if (type_maillage == MED_STRUCTURE) {
      if (MEDnatureGrilleLire(fid,maa,&type) < 0) {
	MESSAGE("Erreur a la lecture de la nature d'une grille");
	return -1;
      }   
      if (type == MED_GRILLE_CARTESIENNE)
	printf("- Grille cartesienne \n");
      if (type == MED_GRILLE_STANDARD)
	printf("- Grille de-structureee \n");
    }
    
    /* On regarde les coordonnees de la grille standard */
    if (type_maillage == MED_STRUCTURE && type == MED_GRILLE_STANDARD) {
      
      nnoeuds = MEDnEntMaa(fid,maa,MED_COOR,MED_NOEUD,0,0);
      if (nnoeuds < 0) {
	MESSAGE("Erreur a la lecture du nombre de noeuds");
	return -1;
      }   
      printf("Nombre de noeuds : "IFORMAT" \n",nnoeuds);
      
      structure_grille = (med_int *) malloc(sizeof(med_int)*mdim);
      if (MEDstructureCoordLire(fid,maa,mdim,structure_grille) < 0) {
	MESSAGE("Erreur a la lecture de la structure de la grille");
	ret = -1;
      }   
      if (ret == 0) {
	printf("Structure des noeuds de la grille : [ ");
	for (j=0;j<mdim;j++)
	  printf(" "IFORMAT" ",*(structure_grille+j));
	printf(" ] \n");
	free(structure_grille);
      }
      
      if (ret == 0) {
	coo = (med_float *) malloc(sizeof(med_float)*nnoeuds*mdim);
	comp2 = (char *) malloc(sizeof(char)*(mdim*MED_TAILLE_PNOM+1));
	unit2 = (char *) malloc(sizeof(char)*(mdim*MED_TAILLE_PNOM+1));

	if ((ret = MEDcoordLire(fid,maa,mdim,coo,MED_FULL_INTERLACE,MED_ALL,NULL,0,
				&repere,comp2,unit2)) < 0) {
	  MESSAGE("Erreur a la lecture des coordonnees des noeuds");
	  ret = -1;
	}   
      }
	
      if (ret == 0) {
	printf("Coordonnees : [ ");
	for (j=0;j<nnoeuds*mdim;j++)
	  printf(" %f ",*(coo+j));
	printf(" ] \n");
	free(coo);
	free(comp2);
	free(unit2);
      }
    } 

    /* On regarde les coordonnees des indices de la grille cartesienne */
    if (type_maillage == MED_STRUCTURE && type == MED_GRILLE_CARTESIENNE) 
      for(axe=1;axe<=mdim;axe++) {
	switch(axe) {
	case 1 :
	  quoi = MED_COOR_IND1;
	  break;
	  
	case 2 :
	  quoi = MED_COOR_IND2;
	  break;
	  
	case 3 :
	  quoi = MED_COOR_IND3;
	  break;
	  
	default :
	  return -1;
	}
	
	nind = MEDnEntMaa(fid,maa,quoi,MED_NOEUD,0,0);
	if (nind < 0) {
	  MESSAGE("Erreur a la lecture de la taille de l'indice");
	  return -1;
	}   
	printf("Lecture de la taille de l'indice : "IFORMAT" \n",nind);
	
	  /* on lit le tableau des indices */
	indices = (med_float *) malloc(sizeof(med_float)*nind);
	if (MEDindicesCoordLire(fid,maa,mdim,indices,nind,axe,comp,unit) < 0) {
	  MESSAGE("Erreur a lecture de indices de coordonnees");
	  ret = -1;
	}   
	if (ret == 0) {
	  printf("Axe %s [%s] : [ ",comp,unit);
	  for (j=0;j<nind;j++)
	    printf(" %f ",indices[j]);
	  printf(" ] \n");
	  free(indices);
	}
      }
  }

  /* On ferme le fichier */
  if (MEDfermer(fid) < 0) {
    MESSAGE("Erreur a la fermeture du fichier");
    return -1;
  }
  printf("Fermeture du fichier \n");
  
  return ret;
}

File: optionals/med-4.0.0/tests/f/2.3.6/test28.f
C*  This file is part of MED.
C*
C*  COPYRIGHT (C) 1999 - 2019  EDF R&D, CEA/DEN
C*  MED is free software: you can redistribute it and/or modify
C*  it under the terms of the GNU Lesser General Public License as published by
C*  the Free Software Foundation, either version 3 of the License, or
C*  (at your option) any later version.
C*
C*  MED is distributed in the hope that it will be useful,
C*  but WITHOUT ANY WARRANTY; without even the implied warranty of
C*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
C*  GNU Lesser General Public License for more details.
C*
C*  You should have received a copy of the GNU Lesser General Public License
C*  along with MED.  If not, see <http://www.gnu.org/licenses/>.
C*

C       ******************************************************************************
C       * - Nom du fichier : test28.f
C       *
C       * - Description : lecture des maillages structures (grille cartesienne |
C       *                 grille de-structuree ) dans le fichier test27.med
C       *
C       *****************************************************************************
	program test28
C       
	implicit none
	include 'med.hf'
C       
C       
	integer*8     fid
	integer       cret,i,j
C       ** la dimension du maillage                         **
	integer       mdim,nind,nmaa,type,quoi,rep,typmaa
C       ** nom du maillage de longueur maxi MED_TAILLE_NOM  **
	character*32  maa
C       ** le nombre de noeuds                              **
	integer       nnoe 
C       ** table des coordonnees                            **
        real*8        coo(8)
	character*16  comp, comp2(2)
	character*16  unit, unit2(2)
	character*200 desc
        integer       strgri(2)
C       ** grille cartesienne                               **
	integer       axe
        real*8        indice(4)
	integer tmp
        
C
C       On ouvre le fichier test27.med en lecture seule
	call efouvr(fid,'test27.med',MED_LECTURE, cret)
	if (cret .ne. 0 ) then
	   print *,'Erreur ouverture du fichier'
	   call efexit(-1)
	endif      
	print *,cret

	print *,'Ouverture du fichier test27.med'
C	
C       Combien de maillage ?
	call efnmaa(fid,nmaa,cret)
	print *,cret
	if (cret .ne. 0 ) then
	   print *,'Erreur lecture du nombre de maillage'
	   call efexit(-1)
	endif      
C
C       On boucle sur les maillages et on ne lit que les
C       maillages structures
	do 10 i=1,nmaa
C
C          On repere les maillages qui nous interessent
C
	   call efmaai(fid,i,maa,mdim,typmaa,desc,cret)
	   print *,cret
	   if (cret .ne. 0 ) then
	      print *,'Erreur lecture maillage info'
	      call efexit(-1)
	   endif      
	   print *,'Maillge de nom : ',maa
	   print *,'- Dimension : ',mdim
	   if (typmaa.eq.MED_STRUCTURE) then
	      print *,'- Type : MED_STRUCTURE'
	   else
	      print *,'- Type : MED_NON_STRUCTURE'   
	   endif
C       
C          On repere le type de la grille
	   if (typmaa.eq.MED_STRUCTURE) then
	      call efnagl(fid,maa,type,cret)
	      print *,cret
	      if (cret .ne. 0 ) then
		 print *,'Erreur lecture nature de la grille'
		 call efexit(-1)
	      endif      
	      if (type.eq.MED_GRILLE_CARTESIENNE) then
		 print *,'- Nature de la grille :',
     &                   'MED_GRILLE_CARTESIENNE'
	      endif
	      if (type.eq.MED_GRILLE_STANDARD) then
		 print *,'- Nature de la grille : MED_GRILLE_STANDARD'
	      endif
	   endif
C
C          On regarde la structure et les coordonnees de la grille MED_GRILLE_STANDARD
	   if ((type.eq.MED_GRILLE_STANDARD) 
     &           .and. (typmaa.eq.MED_STRUCTURE)) then
C
	      call efnema(fid,maa,MED_COOR,MED_NOEUD,0,0,nnoe,cret)
	      print *,cret
	      if (cret .ne. 0 ) then
		 print *,'Erreur lecture nombre de noeud'
		 call efexit(-1)
	      endif      
	      print *,'- Nombre de noeuds : ',nnoe
C
	      call efscol(fid,maa,mdim,strgri,cret)
	      print *,cret
	      if (cret .ne. 0 ) then
		 print *,'Erreur lecture structure de la grille'
		 call efexit(-1)
	      endif      
	      print *,'- Structure de la grille : ',strgri
C
	      call efcool(fid,maa,mdim,coo,
     &                        MED_FULL_INTERLACE,MED_ALL,tmp,
     &                        0,rep,comp2,unit2,cret)
	      print *,cret
	      if (cret .ne. 0 ) then
		 print *,'Erreur lecture des coordonnees des noeuds'
		 call efexit(-1)
	      endif      
	      print *,'- Coordonnees :'
	      do 20 j=1,nnoe*mdim
		 print *,coo(j)
 20	      continue
	   endif
C
	   if ((type.eq.MED_GRILLE_CARTESIENNE)
     &          .and. (typmaa.eq.MED_STRUCTURE)) then
C
	      do 30 axe=1,mdim
		 if (axe.eq.1) then
		    quoi = MED_COOR_IND1
		 endif
		 if (axe.eq.2) then
		    quoi = MED_COOR_IND2
		 endif
		 if (axe.eq.3) then
		    quoi = MED_COOR_IND3
		 endif
C                Lecture de la taille de l'indice selon la dimension
C                fournie par le parametre quoi
		 call efnema(fid,maa,quoi,MED_NOEUD,0,0,nind,cret)
		 print *,cret
		 if (cret .ne. 0 ) then
		    print *,'Erreur lecture taille indice'
		    call efexit(-1)
		 endif      
		 print *,'- Axe ',axe
		 print *,'- Nombre d indices : ',nind
C                Lecture des indices des coordonnees de la grille
		 call eficol(fid,maa,mdim,indice,nind,axe,comp,unit,
     &                       cret)
		 print *,cret
		 if (cret .ne. 0 ) then
		    print *,'Erreur lecture indices de coordonnées'
		    call efexit(-1)
		 endif      
		 print *,'- Axe ',comp
		 print *,'  unite : ',unit
		 do 40 j=1,nind
		    print *,indice(j)
 40		 continue
 30	      continue
C
	   endif
C
 10	continue
C
C       On ferme le fichier
	call efferm (fid,cret)
	print *,cret
	if (cret .ne. 0 ) then
	   print *,'Erreur fermeture du fichier'
	   call efexit(-1)
	endif      
	print *,'Fermeture du fichier'
C
	end
	
